third_party/amd/lib/TritonAMDGPUToLLVM/DotOpToLLVM/MFMA.cpp

```diff
@@ -430,6 +460,284 @@ struct DotOpMFMAConversionHelper {
+              }
+            }
+            auto linearIdx = b * numRepM * numRepN * elemsPerVec +
+                             m * numRepN * elemsPerVec + n * elemsPerVec + v;
+            fc[linearIdx] = accElem;
+          }
+        }
+      }
+    }
+
+    // Originally, setprio (high) is set to the high-level dot op. After dot is
+    // being lowered to the series of mfma operations, it should be moved next
+    // to the first mfma leaving the first mfma staying at the low priority. In
+    // this way, incoming warp can be effectively waiting on the first mfma
+    // instruction (low priority) while the other warp is executing mfma with
+    // high priority. Otherwise, incoming warp can break the cluster.
+    if (setPrioOp && firstMfma)
+      setPrioOp->moveAfter(firstMfma.getDefiningOp());
+
+    // replace with new packed result
+    Type structTy = LLVM::LLVMStructType::getLiteral(
+        ctx, SmallVector<Type>(fc.size(), dstElemTy));
+    Value res = packLLElements(loc, typeConverter, fc, rewriter, structTy);
+
+    Type elemtTy = elemTyA;
+    const size_t mmaCount =
+        numRepB * numRepM * numRepN * numRepK * kWidth / kBase;
+    setNumGeneratedMMAs(op, mmaCount, maybeMfmaInsn->getMDim(),
+                        maybeMfmaInsn->getNDim(), maybeMfmaInsn->getKDim(),
+                        elemtTy);
+
+    rewriter.replaceOp(op, res);
+
+    return success();
+  }
+
+  /// Extract vector from rawElems based on kWidth and kBase
+  /// rawElems is a vector of kWidth elements. We need to prepare vector(s) of
+  /// kBase elements for each mfma instruction
+  SmallVector<Value> extractOperands(Value rawElems, int kWidth, int kBase,
+                                     Type type) const {
+    auto tb = TritonLLVMOpBuilder(loc, rewriter);
+    int kpack = kWidth / kBase;
+    SmallVector<Value> results;
+    auto vecTy = vec_ty(type, kBase);
+    if (type.isBF16())
+      vecTy = vec_ty(i16_ty, kBase);
+    for (int k = 0; k < kpack; ++k) {
+      Value vec = tb.undef(vecTy);
+      for (int elemId = 0; elemId < kBase; ++elemId) {
+        auto val =
+            tb.extract_element(type, rawElems, tb.i32_val(elemId + k * kBase));
+        if (type.isBF16()) {
+          // rocdl.mfma.f32.32x32x8bf16.1k calls for input of i16 type
+          auto cast = tb.bitcast(val, i16_ty);
+          vec = tb.insert_element(vecTy, vec, cast, tb.i32_val(elemId));
+        } else {
+          vec = tb.insert_element(vecTy, vec, val, tb.i32_val(elemId));
+        }
+      }
+      if (type.getIntOrFloatBitWidth() == 8) {
+        if (1 == kBase)
+          results.push_back(tb.zext(i32_ty, tb.bitcast(vec, i8_ty)));
+        if (4 == kBase)
+          // This is for int8 on pre- MI300 GPUs
+          results.push_back(tb.bitcast(vec, i32_ty));
+        if (8 == kBase)
+          results.push_back(tb.bitcast(vec, i64_ty));
+        if (16 == kBase)
+          results.push_back(tb.bitcast(vec, vec_ty(i32_ty, 4)));
```

Comment:
This is just the same as the normal dot's impl other than this part? I think we can just extend the normal dot's impl and extract them to be commonly used by both? It should be fine to add `kBase` 1/16 cases there given the normal dot won't hit it right @zhanglx13?

Reply:
Yes, and also the `1 == kBase` branch